File: prep/implementation/iterations/iteration_cpp/segment_tree_prefix_sum.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

void build(vector<int> &tree, int node, int start, int end,
           const vector<int> &a) {
  if (start == end) {
    tree[node] = a[start];
  } else {
    int mid = start + (end - start) / 2;
    build(tree, 2 * node + 1, start, mid, a);
    build(tree, 2 * node + 2, mid + 1, end, a);
    tree[node] = tree[2 * node + 1] + tree[2 * node + 2];
  }
}

int query(const vector<int> &tree, int node, int start, int end, int l, int r) {
  if (l > end || r < start) {
    return 0;
  }

  if (l <= start && end <= r) {
    return tree[node];
  }

  int mid = start + (end - start) / 2;
  int lval = query(tree, 2 * node + 1, start, mid, l, r);
  int rval = query(tree, 2 * node + 2, mid + 1, end, l, r);

  return lval + rval;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int n, nq, l, r;
  cin >> n;

  vector<int> a(n), b(n);
  vector<int> tree(4 * n);

  for (int i = 0; i < n; i++) {
    cin >> a[i];
    b[i] = a[i];
  }

  for (int i = 1; i < n; i++) {
    b[i] += b[i - 1];
  }

  fill(tree.begin(), tree.end(), 0);

  build(tree, 0, 0, n - 1, a);

  cin >> nq;
  while (nq--) {
    cin >> l >> r;
    int ps = (l == 0 ? b[r] : b[r] - b[l - 1]);
    int ss = query(tree, 0, 0, n - 1, l, r);

    cout << ps << " " << ss << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/main_iterations/iteration5/SegmentTreePrefixSum.java
import java.util.*;

public class SegmentTreePrefixSum {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int n = in.nextInt();
      int[] a = new int[n];
      int[] b = new int[n];
      int[] tree = new int[4 * n];

      for (int i = 0; i < n; i++) {
        a[i] = b[i] = in.nextInt();
      }

      for (int i = 1; i < n; i++) {
        b[i] += b[i - 1];
      }

      build(tree, 0, 0, n - 1, a);

      int nq = in.nextInt();
      while (nq-- > 0) {
        int l = in.nextInt();
        int r = in.nextInt();

        int psum1 = l == 0 ? b[r] : b[r] - b[l - 1];
        int psum2 = query(tree, 0, 0, n - 1, l, r);
        System.out.printf("%d %d\n", psum1, psum2);
      }
    }
  }

  private static void build(int[] tree, int node, int start, int end, int[] a) {
    if (start == end) {
      tree[node] = a[start];
    } else {
      int mid = start + (end - start) / 2;
      build(tree, 2 * node + 1, start, mid, a);
      build(tree, 2 * node + 2, mid + 1, end, a);
      tree[node] = tree[2 * node + 1] + tree[2 * node + 2];
    }
  }

  private static void update(int[] tree, int node, int start, int end, int idx,
                             int val, int[] a) {
    if (start == end) {
      tree[node] = val;
      a[idx] = val;
    } else {
      int mid = start + (end - start) / 2;
      if (idx >= start && idx <= mid) {
        update(tree, 2 * node + 1, start, mid, idx, val, a);
      } else {
        update(tree, 2 * node + 2, mid + 1, end, idx, val, a);
      }

      tree[node] = tree[2 * node + 1] + tree[2 * node + 2];
    }
  }

  private static int query(int[] tree, int node, int start, int end, int l,
                           int r) {
    if (r < start || l > end) {
      return 0;
    }

    if (l <= start && end <= r) {
      return tree[node];
    }

    int mid = start + (end - start) / 2;
    int lval = query(tree, 2 * node + 1, start, mid, l, r);
    int rval = query(tree, 2 * node + 2, mid + 1, end, l, r);

    return lval + rval;
  }
}